core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala

```diff
@@ -17,76 +17,175 @@
+    driverRef.set(ref)
   }
 
-  private def getDelegationTokenProviders: Map[String, HadoopDelegationTokenProvider] = {
-    val providers = Seq(new HadoopFSDelegationTokenProvider(fileSystems)) ++
-      safeCreateProvider(new HiveDelegationTokenProvider) ++
-      safeCreateProvider(new HBaseDelegationTokenProvider)
+  /**
+   * Start the token renewer. Upon start, if a principal has been configured, the renewer will:
+   *
+   * - log in the configured principal, and set up a task to keep that user's ticket renewed
+   * - obtain delegation tokens from all available providers
+   * - schedule a periodic task to update the tokens when needed.
+   *
+   * When token renewal is not enabled, this method will not start any periodic tasks. Instead, it
+   * will generate tokens if the driver ref has been provided, update the current user, and send
+   * those tokens to the driver. No future tokens will be generated, so when that initial set
+   * expires, the app will stop working.
+   *
+   * @param driver If provided, the driver where to send the newly generated tokens.
+   *               The same ref will also receive future token updates unless overridden later.
+   * @return The newly logged in user, or null if a principal is not configured.
+   */
+  def start(driver: Option[RpcEndpointRef] = None): UserGroupInformation = {
+    driver.foreach(setDriverRef)
 
-    // Filter out providers for which spark.security.credentials.{service}.enabled is false.
-    providers
-      .filter { p => isServiceEnabled(p.serviceName) }
-      .map { p => (p.serviceName, p) }
-      .toMap
+    if (principal != null) {
+      renewalExecutor =
+        ThreadUtils.newDaemonSingleThreadScheduledExecutor("Credential Renewal Thread")
+
+      val originalCreds = UserGroupInformation.getCurrentUser().getCredentials()
+      val ugi = doLogin()
+
+      val tgtRenewalTask = new Runnable() {
+        override def run(): Unit = {
+          ugi.checkTGTAndReloginFromKeytab()
+        }
+      }
+      val tgtRenewalPeriod = sparkConf.get(KERBEROS_RELOGIN_PERIOD)
+      renewalExecutor.scheduleAtFixedRate(tgtRenewalTask, tgtRenewalPeriod, tgtRenewalPeriod,
+        TimeUnit.SECONDS)
+
+      val creds = obtainTokensAndScheduleRenewal(ugi)
+      ugi.addCredentials(creds)
+
+      val driver = driverRef.get()
+      if (driver != null) {
+        val tokens = SparkHadoopUtil.get.serialize(creds)
+        driver.send(UpdateDelegationTokens(tokens))
+      }
+
+      // Transfer the original user's tokens to the new user, since it may contain needed tokens
+      // (such as those user to connect to YARN). Explicitly avoid overwriting tokens that already
+      // exist in the current user's credentials, since those were freshly obtained above
+      // (see SPARK-23361).
+      val existing = ugi.getCredentials()
+      existing.mergeAll(originalCreds)
+      ugi.addCredentials(existing)
+      ugi
+    } else {
```

Comment:
Can we make this method overrideable since, despite these two cases being the only ones in Yarn and Mesos, Kubernetes does have a third option with pre-defined secrets, using Kubernetes specific Watchers, I wish to leverage that will also end with `ref.send(UpdateDelegationTokens(tokens)`? Unless you prefer to have the follow-up PR do this refactoring.

Reply:
Yes, I'd prefer any k8s-specific enhancements to be made together with the corresponding k8s code.